File: halt.hs
#!/usr/bin/env runhaskell

--function f
f :: Bool -> Bool
f x = g x

--function g
g :: Bool -> Bool
g x = not x

--main function
main :: IO ()
main = do
--Preamble
  putStrLn ""
  putStrLn "-------------------------------------------------------"
  putStrLn "-- Function f(x) predicts what function g(x) will do --"
  putStrLn "-- given f(g(x)) and then g(x) does the opposite.    --"
  putStrLn "-------------------------------------------------------"
  putStrLn ""
--First input
  putStrLn "The input x = True"
  putStrLn ("f(g(True)) = " ++ (show (f (g True))))
  putStrLn ("g(True) = " ++ (show (g True)))
  putStrLn ""
--Second input
  putStrLn "The input x = False"
  putStrLn ("f(g(False)) = " ++ (show (f (g False))))
  putStrLn ("g(False) = " ++ (show (g False)))
  putStrLn ""

File: halt.cpp
#include <iostream>
using namespace std;

bool f(bool);
bool g(bool);

int main() {
        //Preamble
        cout << endl;
        cout << "-------------------------------------------------------" << endl;
        cout << "-- Function f(x) predicts what function g(x) will do --" << endl;
        cout << "-- given f(g(x)) and then g(x) does the opposite.    --" << endl;
        cout << "-------------------------------------------------------" << endl;
        cout << endl;

        //First input
        cout << "The input x = true" << endl;
        cout << "f(g(true)) = ";
        bool result = f(g(true));
        cout << boolalpha << result << endl;
        cout << "g(true) = ";
        result = g(true);
        cout << result << endl;
        cout << endl;

        //Second input
        cout << "The input x = false" << endl;
        cout << "f(g(false)) = ";
        result = f(g(false));
        cout << result << endl;
        cout << "g(false) = ";
        result = g(false);
        cout << result << endl;
        cout << endl;

        return 0;
}

bool f(bool x) {
        return g(x);
}

bool g(bool x) {
        return !x;
}
